IdentityCore/src/cache/mac/MSIDMacKeychainTokenCache.m

```diff
@@ -443,44 +567,135 @@ - (MSIDCredentialCacheItem *)tokenWithKey:(__unused MSIDCacheKey *)key
 
 // Read one or more credentials from the keychain that match the key (see credentialItem:matchesKey).
 // If not found, return an empty list without setting an error.
-- (NSArray<MSIDCredentialCacheItem *> *)tokensWithKey:(__unused MSIDCacheKey *)key
-                                           serializer:(__unused id<MSIDCacheItemSerializing>)serializer
-                                              context:(__unused id<MSIDRequestContext>)context
-                                                error:(__unused NSError **)error
+- (NSArray<MSIDCredentialCacheItem *> *)tokensWithKey:(MSIDCacheKey *)key
+                                           serializer:(id<MSIDCacheItemSerializing>)serializer
+                                              context:(id<MSIDRequestContext>)context
+                                                error:(NSError **)error
 {
-    NSArray *tokenItems = [self itemsWithKey:key context:context error:error];
-
-    if (!tokenItems)
+    NSArray<MSIDCredentialCacheItem *> *tokenList = [[NSArray alloc] init];
+    
+    if ([key isKindOfClass:([MSIDDefaultCredentialCacheKey class])])
     {
-        return nil;
+        /*
+         Lazy load app and shared credential blob to sync in memory cache with persistent cache.
+         */
+        [self initializeCredentialStorageItem:key serializer:serializer context:context error:error];
+        
+        if (key.isShared)
+        {
+            tokenList = [self.sharedStorageItem storedCredentialsForKey:(MSIDDefaultCredentialCacheKey *)key];
+        }
+        else
+        {
+            tokenList = [self.appStorageItem storedCredentialsForKey:(MSIDDefaultCredentialCacheKey *)key];
+        }
     }
+    
+    return tokenList;
+}
 
-    NSMutableArray<MSIDCredentialCacheItem *> *tokenList = [NSMutableArray new];
-    for (NSDictionary *dict in tokenItems)
+- (void)initializeCredentialStorageItem:(MSIDCacheKey *)key
+                             serializer:(id<MSIDCacheItemSerializing>)serializer
+                                context:(id<MSIDRequestContext>)context
+                                  error:(NSError **)error
+{
+    @synchronized (self)
     {
-        NSData *jsonData = dict[(id)kSecValueData];
-        if (jsonData)
+        MSIDMacCredentialStorageItem *storageItem = nil;
+        
+        if (!key.isShared && ![self.appStorageItem storedCredentialsCount])
         {
-            MSIDCredentialCacheItem *cacheItem = (MSIDCredentialCacheItem *)[serializer deserializeCredentialCacheItem:jsonData];
-            if (cacheItem != nil)
+            storageItem = [self storageItemWithKey:key serializer:serializer context:context error:error];
```

Comment:
some of the logic seems duped. Can we optimize? Also see the comment above that even if store is non empty, we should still read keychain in certain cases.